code/game/objects/items/stacks/stack.dm

```diff
@@ -227,26 +227,36 @@
 		else
 			to_chat(usr, "<span class='warning'>You haven't got enough [src] to build \the [R.title]!</span>")
 		return FALSE
-	if(R.window_checks && !valid_window_location(usr.loc, usr.dir))
+	var/turf/T = get_turf(usr)
+	if(R.window_checks && !valid_window_location(T, usr.dir))
 		to_chat(usr, "<span class='warning'>The [R.title] won't fit here!</span>")
 		return FALSE
-	if(R.one_per_turf && (locate(R.result_type) in usr.loc))
+	if(R.one_per_turf && (locate(R.result_type) in T))
 		to_chat(usr, "<span class='warning'>There is another [R.title] here!</span>")
 		return FALSE
-	if(R.on_floor && !isfloorturf(usr.loc))
-		to_chat(usr, "<span class='warning'>\The [R.title] must be constructed on the floor!</span>")
-		return FALSE
+	if(R.on_floor)
+		if(!isfloorturf(T))
+			to_chat(usr, "<span class='warning'>\The [R.title] must be constructed on the floor!</span>")
+			return FALSE
+		for(var/obj/AM in T)
+			if(istype(AM,/obj/structure/grille))
+				continue
+			if(istype(AM,/obj/structure/table))
+				continue
+			if(AM.density)
+				to_chat(usr, "<span class='warning'>There is a [AM.name] blocking you cant fit a [R.title] here!</span>")
```

Comment:
You might want to reword this

Reply:
Please suggest a wording that is ambigous for any structure, machine, closed airlock or other object with a density of one.